compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs

```diff
@@ -164,44 +164,48 @@ fn is_valid_cmse_output<'tcx>(
 ) -> Result<bool, &'tcx LayoutError<'tcx>> {
     // this type is only used for layout computation, which does not rely on regions
     let fn_sig = tcx.instantiate_bound_regions_with_erased(fn_sig);
+    let fn_sig = tcx.erase_and_anonymize_regions(fn_sig);
+    let return_type = fn_sig.output();
+
+    // `impl Trait` is already disallowed with `cmse-nonsecure-call`, because that ABI is only
+    // allowed on function pointers, and function pointers cannot contain `impl Trait` in their
+    // signature.
+    //
+    // Here we explicitly disallow `impl Trait` in the `cmse-nonsecure-entry` return type too, to
+    // prevent query cycles when calculating the layout. This ABI is meant to be used with
+    // `#[no_mangle]` or similar, so generics in the type really don't make sense.
```

Comment:
for the record opaque types are not generics. `fn foo() -> impl Trait`'s return type is some fully concrete non generic type, its just not explicitly written out. I think you're right that `no_mangle`'d functions with RPITs are silly though

Reply:
really? My intuition is that `-> impl Trait` is polymorphic and that it is roughly equivalent to `fn foo<T: Trait>() -> T`. You'd need to pick a concrete imlementation of the trait in order to actually be able to calculate the layout of the return type (and from that the ABI).

Is there some better word for "the type is not known enough to calculate the layout"?

Reply:
the type of RPITs *should* be known enough to calculate the layout of the return type as long as opaque types are able to be properly normalized. `impl Trait` only desugars to a generic parameter in *argument* position, i.e. `fn foo(a: impl Trait)` desugars as you say

Reply:
Oh interesting, it actually looks at the body to figure out the concrete type if it occurs just in return position

https://godbolt.org/z/KhfhrxPG6

Reply:
But at the point where we attempt to generate the layout for these entry functions, somehow there is a loop in the calculation

https://godbolt.org/z/K58sMorf3

(that godbolt only works with current nightly, but the same cycle is in the issue https://github.com/rust-lang/rust/issues/147242)

anyway we're totally fine with just disallowing `impl Trait` entirely, it's a simple rule that doesn't really limit any practical use.

Reply:
yeah I would be interested in seeing where the cycle comes from, but it does seem fine to just altogether forbid this as RPITs in no_mangle functions seems like a big footgun/code smell